### PR TITLE
OAuth over SASL_PLAIN

### PR DESCRIPTION
**Pull request by mstruk**
OAuth over PLAIN

An implementation of using SASL_PLAIN to authenticate with a clientId and a secret or with an access token.
See README.md for instructions on how to set it up.

### CODE REVIEW

File: examples/producer/src/main/java/io/strimzi/examples/producer/ExampleProducer.java
Comment (scholzj) [inline examples/producer/src/main/java/io/strimzi/examples/producer/ExampleProducer.java:123]: Is it intentional to have single example with both OAUTHBEARER and PLAIN in one like this? If yes, it might be nice to add some explanatory comments here in the code to explain what and when should be un-commented and what should be commented.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/common/LogUtil.java
Comment (scholzj) [inline oauth-common/src/main/java/io/strimzi/kafka/oauth/common/LogUtil.java:45]: Should it be maybe named in plural? `getCauseMessages`? Or maybe even `getAllCauseMessages` to make this a bit more self-explaining?

File: oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java
Comment (scholzj) [inline oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java:106]: I wonder if we should give this name a bit more special feeling. for example `$accessToken` This fits with what for example Microsoft does on EventHubs. I think it makes it a bit more clear that it is some special value with special meaning when reading the different configuration files using it.

**Review by scholzj — COMMENTED** [review]
Some nits, but LGTM otherwise.

File: oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java
Comment (mstruk) [reply oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java:106]: Makes sense, but for this specific suggestion of `$accessToken` my concern is that it would cause problems in the CLI use-cases. E.g.:

    kafkacat -b my-cluster-kafka-bootstrap:9092 -X security.protocol=SASL_PLAINTEXT -X sasl.mechanism=PLAIN -X sasl.username=$accessToken -X sasl.password="$ACCESS_TOKEN" -t a_topic -P

This will result in empty username, and authentication will fail. For that reason I would not use $ in the reserved word.
But maybe we could use @ as in `@accessToken`.

Comment (scholzj) [reply oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java:106]: MS is using `$` ... that was why I suggested it. But you have a good point, although I'm not sure how often will people really use it from the command line. 

@strimzi/maintainers WDYT?.

File: examples/producer/src/main/java/io/strimzi/examples/producer/ExampleProducer.java
Comment (mstruk) [reply examples/producer/src/main/java/io/strimzi/examples/producer/ExampleProducer.java:123]: Good catch, this was not intended this way.

File: oauth-common/src/main/java/io/strimzi/kafka/oauth/common/LogUtil.java
Comment (mstruk) [reply oauth-common/src/main/java/io/strimzi/kafka/oauth/common/LogUtil.java:45]: Makes sense.

File: oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java
Comment (tombentley) [inline oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java:31]: Javadoc

File: oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java
Comment (tombentley) [reply oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java:106]: I don't have a strong opinion. If it's open to interpolation by the shell it can be enclosed in `'`.

File: oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipal.java
Comment (tombentley) [inline oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipal.java:48]: If these don't support stack-like push and pop semantics then are the method names really the right ones?

File: README.md
Comment (mstruk) [reply README.md:386]: It refers to the re-authentication mechanism which only exists to support SASL mechanisms that are token based, and require token refresh in order for the client to be able to keep using the current connection. When you authenticate with SASL_OAUTHBEARER the Kafka broker will go through a dialog with the client when the token expires. It won't ever enter that dialog when you authenticate with SASL_PLAIN and the token expires. In the latter case any additional interaction will simply result in the `AuthorizationException`.

Comment (mstruk) [reply README.md:386]: The part about being disabled by default refers to the fact that unless you specify `maxSecondsWithoutReauthentication` config option the re-authentication won't even be activated for SASL_OAUTHBEARER.

File: oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipal.java
Comment (mstruk) [reply oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipal.java:48]: Fair point. Will rename.

File: oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java
Comment (mstruk) [reply oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java]: Ok, so I guess '$accessToken' seems more natural to most people, and they would understand this as setting a kind of server side variable rather than refer to the client side environment variable. And escaping it in the shell is simple.

File: README.md
Comment (mstruk) [reply README.md]: I reworded it to make it more clear (hopefully).

File: oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java
Comment (mstruk) [reply oauth-server-plain/src/main/java/io/strimzi/kafka/oauth/server/plain/JaasServerOauthOverPlainValidatorCallbackHandler.java:98]: Added Javadoc, and to [JaasServerOauthValidatorCallbackHandler](https://github.com/strimzi/strimzi-kafka-oauth/pull/77/files#diff-0af30cbf06d7deee32bafafb11afc8d0f5ac5242878a6e7746f4f315ca94ea11) as well.

**Review by tombentley — APPROVED** [review]

**Review by scholzj — APPROVED** [review]

**Comment by scholzj** [comment]
@mstruk Can this be merged? Should we do 0.7.0-rc1 after merging this or will there be some more work needed for 0.7.0?

**Comment by mstruk** [comment]
@scholzj It is ready for merge.

I'm working on another PR for 0.7.0 - a functionality to allow enforcing the audience - a mechanism where the token is specifically issued for use by a limited set of resource servers. That makes it possible for Kafka broker to reject the tokens not issued for it during authentication already. While KeycloakRBACAuthorizer performs all the necessary checks during each authorization, one might want to turn off authorization (maybe during development) and still limit the access by only allowing some accounts to authenticate.

**Comment by scholzj** [comment]
@mstruk Ok, thanks. I will wait with the RC then.

**Pull request merged**
